### PR TITLE
Add Priority Hints IDL and preference

### PR DESCRIPTION
**Pull request by rwlbuis**
#### adf13d454519056c7aa9dad15907ef23263f204e
<pre>
Add Priority Hints IDL and preference
<a href="https://bugs.webkit.org/show_bug.cgi?id=247291">https://bugs.webkit.org/show_bug.cgi?id=247291</a>

Reviewed by Youenn Fablet.

Add Priority Hints IDL for img, script and link HTML elements as well as Fetch&apos;s RequestInit.

Hide Priority Hints functionality behind a preference.

This change also adds WPT tests relevant to Priority Hints.

* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/request-init-priority.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/request-init-priority.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/request-init-priority.any.js: Added.
(idx.in.priorities.test):
(test):
(idx.in.priorities.promise_test):
(promise_test):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/request-init-priority.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/request-init-priority.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/request/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/img-attr-fetchpriority.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/img-attr-fetchpriority.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/link-attr-fetchpriority.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/link-attr-fetchpriority.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/link-header-fetchpriority.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/link-header-fetchpriority.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/link-header-fetchpriority.tentative.html.sub.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/resources/script.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/resources/stylesheet.css: Added.
(body):
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/script-attr-fetchpriority.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/script-attr-fetchpriority.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/priority-hints/w3c-import.log: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/fetch/FetchRequestInit.h:
* Source/WebCore/Modules/fetch/FetchRequestInit.idl:
* Source/WebCore/Modules/fetch/RequestPriority.h: Added.
* Source/WebCore/Modules/fetch/RequestPriority.idl: Copied from Source/WebCore/Modules/fetch/FetchRequestInit.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::setFetchPriorityForBindings):
(WebCore::HTMLImageElement::fetchPriorityForBindings const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setFetchPriorityForBindings):
(WebCore::HTMLLinkElement::fetchPriorityForBindings const):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLLinkElement.idl:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::setFetchPriorityForBindings):
(WebCore::HTMLScriptElement::fetchPriorityForBindings const):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.idl:

Canonical link: <a href="https://commits.webkit.org/260666@main">https://commits.webkit.org/260666@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b3b3b3d01d5b8066acfe6ec90047b40337ccca04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41734 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/435 "Archived built product (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118210 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9278 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101136 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14595 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97814 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42702 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96567 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29455 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84437 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97985 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30804 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98776 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8899 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11524 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7736 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30994 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16914 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50400 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106418 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7366 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13112 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26378 "Passed tests") | 
| | | | |

### CODE REVIEW

**Comment by rwlbuis** [comment]
@youennf ping :)

File: Source/WebCore/Modules/fetch/FetchRequestInit.h [hunk 1]
```diff
@@ -47,6 +47,7 @@ struct FetchRequestInit {
     String integrity;
     std::optional<bool> keepalive;
     JSC::JSValue signal;
+    std::optional<FetchPriority> priority;
```
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchRequestInit.h]: optional or default value being `auto`?

File: Source/WebCore/Modules/fetch/FetchRequestInit.idl [hunk 2]
```diff
@@ -41,5 +41,6 @@ dictionary FetchRequestInit {
     boolean keepalive;
     // FIXME: 'signal' should be of type AbortSignal?.
     any signal;
+    RequestPriority priority;
```
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchRequestInit.idl]: Can we add EnabledBySetting=PriorityHintsEnabled?

File: Source/WebCore/Modules/fetch/RequestPriority.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ReferrerPolicy.h"
```
Comment (youennf) [inline Source/WebCore/Modules/fetch/RequestPriority.h]: FetchPriority.h?

File: Source/WebCore/loader/FetchOptions.h [hunk 4]
```diff
@@ -33,6 +33,7 @@
 #include "FetchOptionsDestination.h"
 #include "FetchOptionsMode.h"
 #include "FetchOptionsRedirect.h"
+#include "FetchPriority.h"
```
Comment (youennf) [inline Source/WebCore/loader/FetchOptions.h]: Probably not needed right now.

File: Source/WebCore/loader/FetchPriority.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+enum class FetchPriority : uint8_t { High, Low, Auto };
+
+String fetchPriorityToString(const FetchPriority&);
+FetchPriority parseFetchPriority(StringView);
+
+}
+
+namespace WTF {
+
+template<> struct EnumTraitsForPersistence<WebCore::FetchPriority> {
```
Comment (youennf) [inline Source/WebCore/loader/FetchPriority.h]: Why EnumTraitsForPersistence?
Are we expecting to keep that info in DOMCache?
Maybe we should add it later.

File: Source/WebCore/loader/FetchPriority.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    case FetchPriority::High:
+        return "high"_s;
+    case FetchPriority::Low:
+        return "low"_s;
+    case FetchPriority::Auto:
+        return "auto"_s;
+    }
+    ASSERT_NOT_REACHED();
+    return { };
+}
```
Comment (youennf) [inline Source/WebCore/loader/FetchPriority.cpp]: This code is probably generated by the binding generator.
We could either use RequestPriority everywhere or include JSRequestPriority.h and add the necessary WebIDL key words to generate parsing/serialization of the enum.

**Review by youennf — APPROVED** [review]

File: Source/WebCore/Modules/fetch/FetchRequestInit.idl (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/Modules/fetch/FetchRequestInit.idl]: Done.

File: Source/WebCore/Modules/fetch/RequestPriority.h (hunk 3 above)
Comment (rwlbuis) [reply Source/WebCore/Modules/fetch/RequestPriority.h]: Yes, this is from an earlier version, anyway in the end it is clearer to just only keep RequestPriority.

File: Source/WebCore/loader/FetchOptions.h (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/loader/FetchOptions.h]: Removed.

File: Source/WebCore/loader/FetchPriority.h (hunk 5 above)
Comment (rwlbuis) [reply Source/WebCore/loader/FetchPriority.h]: Copy pasta, removed.

File: Source/WebCore/loader/FetchPriority.cpp (hunk 6 above)
Comment (rwlbuis) [reply Source/WebCore/loader/FetchPriority.cpp]: Done. I could see only one fetch data structure where we did this before, looks like there is room to do more (for ReferrerPolicy the parsing is likely duplicated.

File: Source/WebCore/Modules/fetch/FetchRequestInit.h (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/Modules/fetch/FetchRequestInit.h]: Maybe we can fine tune later, either way is fine to me.

**Pull request merged**